File: QID-1324-SFSbitreeNDiv/SFSbitreeNDiv.R

# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

# install and load packages
libraries = c("fUtilities")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

# parameter settings
t    = 2      # time
n    = 10     # step
u    = 1.1    # upward proportion
d    = 0.9    # downward proportion
S0   = 50     # initial price
type = 1      # 0 for American, 1 for European
k    = 50     # strike price
i    = 0.05   # interest rate
flag = 1      # 0 for put, 1 for call
b    = i      # Costs of carry

# Main Computation
dt = t/n  # Interval of step
p  = (exp(b) - d)/(u - d)  # Pseudo probability of up movement
un = matrix(0, n + 1, 1)
un[n + 1, 1] = 1
dm = t(un)
um = dm
j  = 1

while (j < n + 1) {
    # Down movement dynamics
    d1 = c(matrix(0, 1, n - j), (matrix(1, 1, j + 1) * d)^(seq(0:j) - 1))
    dm = rbind(dm, d1)
    # Up movement dynamics
    u1 = c(matrix(0, 1, n - j), matrix(1, 1, j + 1) * u^seq(j, 0, -1))
    um = rbind(um, u1)
    j  = j + 1
}

um = t(um)
dm = t(dm)

# Stock price development
s = S0 * um * dm
colnames(s) = c()

print("Stock price development")
print(s)

# Rearangement
s   = s[ncol(s):1, ]
opt = matrix(0, ncol(s), ncol(s))

if (flag == 1 & type == 0) {
    # Option is a American call
    opt[, n + 1] = rowMaxs(s[, ncol(s)] - k, 0)  # Determine option values from prices
    for (j in n:1) {
        l = 1:j
        # Probable option values discounted back one time step
        discopt = ((1 - p) * opt[l, j + 1] + p * opt[l + 1, j + 1]) * exp(-b * 
            dt)
        # Option value is max of current stock price - strike or discopt
        opt[, j] = c(rowMaxs(cbind(s[1:j, j] - k, discopt)), matrix(0, n + 1 - 
            j, 1))
    }
    American_Call_Price = opt[ncol(opt):1, ]
    
    print("Option price development")
    print(American_Call_Price)
    
    print("The price of the American call option at time t_0 is")
    print(American_Call_Price[n + 1, 1])
    
}

if (flag == 1 & type == 1) {
    # Option is a European call
    opt[, n + 1] = rowMaxs(s[, ncol(s)] - k, 0)  # Determine option values from prices
    for (j in n:1) {
        l = 1:j
        # Probable option values discounted back one time step
        discopt = ((1 - p) * opt[l, j + 1] + p * opt[l + 1, j + 1]) * exp(-b * 
            dt)
        # Option value is max of current stock price - strike or discopt
        opt[, j] = c(discopt, matrix(0, n + 1 - j, 1))
    }
    
    European_Call_Price = opt[ncol(opt):1, ]
    
    print("Option price development")
    print(European_Call_Price)
    
    print("The price of the European call option at time t_0 is")
    print(European_Call_Price[n + 1, 1])
    
}

if (flag == 0 & type == 0) {
    # Option is an American put
    opt[, n + 1] = rowMaxs(k - s[, ncol(s)], 0)  # Determine option values from prices
    for (j in n:1) {
        l = 1:j
        # Probable option values discounted back one time step
        discopt = ((1 - p) * opt[l, j + 1] + p * opt[l + 1, j + 1]) * exp(-b * 
            dt)
        # Option value is max of current stock price - strike or discopt
        opt[, j] = c(rowMaxs(cbind(k - s[1:j, j], discopt)), matrix(0, n + 1 - 
            j, 1))
    }
    American_Put_Price = opt[ncol(opt):1, ]
    
    print("Option price development")
    print(American_Put_Price)
    
    print("The price of the American put option at time t_0 is")
    print(American_Put_Price[n + 1, 1])
    
}

if (flag == 0 & type == 1) {
    # Option is a European put
    opt[, n + 1] = rowMaxs(k - s[, ncol(s)], 0)  # Determine option values from prices
    for (j in n:1) {
        l = 1:j
        # Probable option values discounted back one time step
        discopt = ((1 - p) * opt[l, j + 1] + p * opt[l + 1, j + 1]) * exp(-b * 
            dt)
        # Option value is max of current stock price - strike or discopt
        opt[, j] = c(discopt, matrix(0, n + 1 - j, 1))
    }
    
    European_Put_Price = opt[ncol(opt):1, ]
    
    print("Option price development")
    print(European_Put_Price)
    
    print("The price of the European put option at time t_0 is")
    print(European_Put_Price[n + 1, 1])
    
}

File: QID-1324-SFSbitreeNDiv/SFSbitreeNDiv.m
% ---------------------------------------------------------------------
% Book:         SFS
% ---------------------------------------------------------------------
% Quantlet:     SFSbitreeNDiv
% ---------------------------------------------------------------------
% Description:  SFSbitreeNDiv computes European/American option prices
%               using a binomial tree for assets without dividends.
% ---------------------------------------------------------------------
% Usage:        SFSbitreeNDiv
% ---------------------------------------------------------------------
% Inputs:       S0 - Stock price
%               k - Exercise price
%               i - Interest rate
%               t - Time to expiration
%               n - Number of intervals
%               type - 0 is American/1 is European
%               flag - 1 is call/0 is put
%               u - Upper proportion
%               d - Downward proportion
% ---------------------------------------------------------------------
% Output:      Binomial trees and price of option
% ---------------------------------------------------------------------
% Example:     Refer to exercise 8.6 in SFS
% ---------------------------------------------------------------------
% Author:      Weining Wang 20090416
% ---------------------------------------------------------------------
%

clear
clc

t    = 2;     % time
n    = 10;    % step
u    = 1.1;   % upper proportion
d    = 0.9;   % downward proportion
S0   = 50;    % initial price
type = 0;     % 0 for American , 1 for European
k    = 50;    % strike price
i    = 0.05;  % interest rate
flag = 1;     % 0 for put, 1 for call
b    = i;     % Costs of carry

%% Main Computation

dt        = t/n;                % Interval of step
p         = (exp(b)-d)/(u-d);   % Pseudo probability of up movement

un        = zeros(n+1,1);
un(n+1,1) = 1;
dm        = un';
um        = [];
j         = 1;

while j < n+1
    d1 = [zeros(1,n-j) (ones(1,j+1)*d).^((1:j+1)-1)];
    dm = [dm; d1];                                       % Down movement dynamics
    u1 = [zeros(1,n-j) (ones(1,j+1)*u).^((j:-1:0))];
    um = [um; u1];                                       % Up movement dynamics
    j  = j+1;
end

um          = [un';um]';
dm          = dm';

s           = S0.*um.*dm;                                % Stock price development

Stock_Price = s
s           = flipud(s);                                 % Rearangement
opt         = zeros(size(s));

if flag == 1 & type == 0 % Option is a American call
    opt(:,n+1) = max(s(:,end)-k,0);  % Determine option values from prices
    for j = n:-1:1;
        l = 1:j;
        % Probable option values discounted back one time step
        discopt = ((1-p)*opt(l,j+1)+p*opt(l+1,j+1))*exp(-b*dt);
        % Option value is max of current stock price - strike or discopt
        opt(:,j) = [max(s(1:j,j)-k,discopt);zeros(n+1-j,1)];
    end
    American_Call_Price = flipud(opt)
    disp(' ') ;
    disp('The price of the American call option at time t_0 is')
    disp(American_Call_Price(n+1,1))
    

elseif flag == 1 & type == 1 % Option is a European call
    opt(:,n+1) = max(s(:,n+1)-k,0);  % Determine option values from prices
    for j = n:-1:1;
        l = 1:j;
        % Probable option values discounted back one time step
        discopt = ((1-p)*opt(l,j+1)+p*opt(l+1,j+1))*exp(-b*dt);
        % Option value
        opt(:,j) = [discopt;zeros(n+1-j,1)];
    end
    European_Call_Price = flipud(opt)
    disp(' ') ;
    disp('The price of the European call option at time t_0 is')
    disp(European_Call_Price(n+1,1))
     

elseif flag == 0 & typ == 0 % Option is an American put                              
    opt(:,n+1) = max(k-s(:,n+1),0);  % Determine option values from prices
    for j = n:-1:1
        l = 1:j;
        % Probable option values discounted back one time step
        discopt = ((1-p)*opt(l,j+1)+p*opt(l+1,j+1))*exp(-b*dt);
        % Option value is max of X - current price or discopt
        opt(:,j) = [max(k-s(1:j,j),discopt);zeros(n+1-j,1)];
        find(max(k-s(1:j,j),0)>discopt);
    end
    American_Put_Price = flipud(opt)
    disp(' ') ;
    disp('The price of the American put option at time t_0 is')
    disp(American_Put_Price(n+1,1))    
    

elseif flag == 0 & type == 1 % Option is a European put                           
    opt(:,n+1) = max(k-s(:,n+1),0);  % Determine option values from prices
    for j = n:-1:1
        l = 1:j;
        % Probable option values discounted back one time step
        discopt = ((1-p)*opt(l,j+1)+p*opt(l+1,j+1))*exp(-b*dt);
        % Option value
        opt(:,j) = [discopt;zeros(n+1-j,1)];
    end
    European_Put_Price = flipud(opt)
    disp(' ') ;
    disp('The price of the European put option at time t_0 is')
    disp(European_Put_Price(n+1,1))
    
end


